lib/reconcile/reconcile.go

```diff
@@ -0,0 +1,200 @@
+type Plan interface {
+	// Name returns the name of this plan.
+	// The name is used in logging
+	Name() string
+	// Create creates a list of operation steps to execute as part of the plan.
+	// Individual steps are capable of generating sub-steps as part of their work.
+	Create(context.Context) ([]Step, error)
+}
+
+// Step describes a single reconciler step
+type Step interface {
+	// Name returns the step name
+	Name() string
+	// Do executes the step.
+	// Optionally returns additional Steps to execute or error in case of failure
+	Do(context.Context) ([]Step, error)
+}
+
+// New creates a new plan reconciler
+func New(config Config) *Reconciler {
+	config.setDefaults()
+	return &Reconciler{
+		config:      config,
+		log:         logrus.WithField(trace.Component, "reconciler"),
+		stop:        make(chan chan struct{}),
+		planUpdates: make(chan Plan, 1),
+	}
+}
+
+// Reset starts the reconciler loop using the specified plan.
+// If the reconciler is already executing a plan, it will be canceled to start a new one
+func (r *Reconciler) Reset(ctx context.Context, plan Plan) {
+	r.once.Do(func() {
+		go r.loop()
+	})
+	select {
+	case <-ctx.Done():
+	case r.planUpdates <- plan:
+	}
+}
+
+// Stop stops the reconciler loop/
+// Stop cannot be invoked multiple times
+func (r *Reconciler) Stop() {
+	c := make(chan struct{})
+	r.stop <- c
+	<-c
+}
+
+// Reconciler is a plan reconciler
+type Reconciler struct {
+	config      Config
+	log         logrus.FieldLogger
+	stop        chan chan struct{}
+	once        sync.Once
+	planUpdates chan Plan
+}
+
+// Config defines reconciler configuration
+type Config struct {
+	// Timeout defines the time steps for executing a single plan between attempts
+	Timeout time.Duration
+	// ResyncTimeout defines the maximum time to wait between unconditional attempts
+	// to execute the last plan
+	ResyncTimeout time.Duration
+	// clock specifies the time implementation.
+	// Overridden in tests
+	clock clockwork.Clock
+}
+
+func (r *Config) setDefaults() {
+	if r.Timeout == 0 {
+		r.Timeout = timeout
+	}
+	if r.ResyncTimeout == 0 {
+		r.ResyncTimeout = resyncTimeout
+	}
+	if r.clock == nil {
+		r.clock = clockwork.NewRealClock()
+	}
+}
+
+func (r *Reconciler) loop() {
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+	for {
+		select {
+		case c := <-r.stop:
+			r.log.Debug("Stop.")
+			r.stop = nil
+			cancel()
+			wg.Wait()
+			close(c)
+			return
+
+		case plan := <-r.planUpdates:
+			cancel()
+			wg.Wait()
+			ctx, cancel = context.WithCancel(context.Background())
+			wg.Add(1)
+			go r.executorLoop(ctx, plan, &wg)
+		}
+	}
+}
+
+func (r *Reconciler) executorLoop(ctx context.Context, plan Plan, wg *sync.WaitGroup) {
+	defer wg.Done()
+	timeout := r.config.Timeout
+	if r.executePlan(ctx, plan) == nil {
+		timeout = r.config.ResyncTimeout
+	}
+	ticker := r.config.clock.NewTicker(timeout)
+	defer ticker.Stop()
+	tickerC := ticker.Chan()
+	for {
```

Comment:
So this seems to be executing the plan forever, not just until success? Is that on purpose? I wonder if there should be some way to pause it (other than stopping planet-agent) when you need to make some changes "out-of-band" to troubleshoot something for example.

Reply:
The purpose of the loop is bringing the system to the desired state hence it is running forever, albeit with longer intervals with no changes to the plan.
I'm not sure about out-of-band changes short of stopping the agent - do kubernetes components (as an example) have an ability to stop their loops for out-of-band troubleshooting?